### PR TITLE
Add disk image transfer metadata form

### PR DESCRIPTION
**Pull request by mistydemeo**
This adds a new metadata input form which is provided when creating disk image transfers.

"Disk image" is exposed as a new transfer type, which operates identically to a standard transfer in every way when processing. When "disk image" is selected, a metadata edit button is added for each transfer component which allows a separate form to be filled out for each component. An edit button is also provided next to the "start" button, which lets the user pre-fill out a row to be attached to the next component that is added.

If a transfer metadata form exists for a given transfer, it will be associated with the "objects" directory of the final AIP produced from that transfer. This is accomplished by serializing it to disk, so it should work with SIP arrange as long as the arranged SIP contains contents from only one disk image transfer.

This adds a few forms to view the metadata form field values; there is code to edit them, but at the moment it's disabled by request.

### CODE REVIEW

File: src/dashboard/src/main/templatetags/keyvalue.py [hunk 1]
```diff
@@ -21,4 +21,7 @@
 
 @register.filter
 def keyvalue(d, key):
-    return d[key]
```
Comment (Hwesta) [inline src/dashboard/src/main/templatetags/keyvalue.py:24]: `return d.get('key')` ?

File: src/MCPServer/share/mysql_dev1.sql [hunk 2]
```diff
@@ -427,3 +427,150 @@ UPDATE MicroServiceChainLinks SET defaultNextChainLink='f2e784a0-356b-4b92-9a5a-
...
+
+ALTER TABLE Transfers ADD COLUMN transferMetadataSetRowUUID VARCHAR(36);
+
+CREATE TABLE TransferMetadataFields (
+  pk varchar(36) NOT NULL,
+  createdTime timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  fieldLabel VARCHAR(50) DEFAULT '',
+  fieldName VARCHAR(50) NOT NULL,
+  fieldType VARCHAR(50) NOT NULL,
+  optionTaxonomyUUID VARCHAR(36) NOT NULL,
```
Comment (Hwesta) [inline src/MCPServer/share/mysql_dev1.sql]: Does this need a foreign key constraint?

File: src/MCPServer/share/mysql_dev1.sql [hunk 3]
```diff
@@ -427,3 +427,150 @@ UPDATE MicroServiceChainLinks SET defaultNextChainLink='f2e784a0-356b-4b92-9a5a-
...
+    VALUES ('0c1b4233-fbe7-463f-8346-be6542574b86', UNIX_TIMESTAMP(), 'Imaging software', 'imaging_software', 'select', 'cae76c7f-4d8e-48ee-9522-4b3fbf492516', 13);
+
+INSERT INTO TransferMetadataFields (pk, createdTime, fieldLabel, fieldName, fieldType, sortOrder)
+    VALUES ('0a9e346a-f08c-4e0d-9753-d9733f7205e5', UNIX_TIMESTAMP(), 'Image fixity', 'image_fixity', 'textarea', 14);
+
+CREATE TABLE TransferMetadataFieldValues (
+  pk varchar(36) NOT NULL,
+  createdTime timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  setUUID VARCHAR(36) NOT NULL,
+  fieldUUID VARCHAR(36) NOT NULL,
```
Comment (Hwesta) [inline src/MCPServer/share/mysql_dev1.sql]: Do these need foreign key constraints?

File: src/MCPServer/share/mysql_dev1.sql [hunk 4]
```diff
@@ -427,3 +427,150 @@ UPDATE MicroServiceChainLinks SET defaultNextChainLink='f2e784a0-356b-4b92-9a5a-
...
+INSERT INTO Taxonomies (pk, name) VALUES ('f6980e68-bac2-46db-842b-da4eba4ba418', 'Disk media densities');
+INSERT INTO Taxonomies (pk, name) VALUES ('31e0bdc9-1114-4427-9f37-ca284577dcac', 'Filesystem types');
+INSERT INTO Taxonomies (pk, name) VALUES ('fbf318db-4908-4971-8273-1094d2ba29a6', 'Disk imaging interfaces');
+INSERT INTO Taxonomies (pk, name) VALUES ('cc4231ef-9886-4722-82ec-917e60d3b2c7', 'Disk image format');
+INSERT INTO Taxonomies (pk, name) VALUES ('cae76c7f-4d8e-48ee-9522-4b3fbf492516', 'Disk imaging software');
+
+CREATE TABLE TaxonomyTerms (
+  pk varchar(36) NOT NULL,
+  createdTime timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  taxonomyUUID varchar(36) NOT NULL,
```
Comment (Hwesta) [inline src/MCPServer/share/mysql_dev1.sql]: Foreign key?

**Comment by mistydemeo** [comment]
Since syncdb now creates the tables, dropped the table definitions and properly defined the foreign keys in the models.

File: src/dashboard/src/main/models.py [hunk 5]
```diff
@@ -219,6 +219,7 @@ class Transfer(models.Model):
     currentlocation = models.TextField(db_column='currentLocation')
     type = models.CharField(max_length=50, db_column='type')
     accessionid = models.TextField(db_column='accessionID')
+    transfermetadatasetrowuuid = models.TextField(db_column='transferMetadataSetRowUUID')
```
Comment (mistydemeo) [inline src/dashboard/src/main/models.py]: Foreign key

File: src/dashboard/src/main/models.py [hunk 6]
```diff
@@ -626,3 +627,51 @@ class AtkDIPObjectResourcePairing(models.Model):
...
+    createdtime = models.DateTimeField(db_column='createdTime', auto_now_add=True)
+    createdbyuserid = models.IntegerField(db_column='createdByUserID')
+
+    class Meta:
+        db_table = u'TransferMetadataSets'
+
+class TransferMetadataField(models.Model):
+    id = UUIDPkField()
+    createdtime = models.DateTimeField(db_column='createdTime', auto_now_add=True)
+    fieldlabel = models.TextField(db_column='fieldLabel')
```
Comment (mistydemeo) [inline src/dashboard/src/main/models.py]: Need to fix types and lengths

**Comment by mistydemeo** [comment]
This PR contains some dead forms. When originally created, this branch had some customizeability in the metadata forms; it was possible to add new taxonomies and terms. It was ultimately decided that the feature as currently built was too specific to the originating client to really justify exposing this via the UI, so the links to the forms were removed from the UI. They're still present in the code in case we decide to hook them up again, though.

**Comment by mistydemeo** [comment]
Pushed a few additional cleanups and some bugfixes. All seems to be good now.

**Comment by mistydemeo** [comment]
Cleaned up commit history.

File: src/dashboard/src/main/templatetags/keyvalue.py (hunk 1 above)
Comment (mistydemeo) [inline src/dashboard/src/main/templatetags/keyvalue.py:24]: Github is still showing this comment here b/c of the diff, but I changed this.

File: src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py [hunk 7]
```diff
@@ -852,6 +904,8 @@ def createFileSec(directoryPath, parentDiv):
     structMap = etree.Element("structMap", TYPE='physical', ID='structMap_1', LABEL="Archivematica default")
     structMapDiv = etree.SubElement(structMap, 'div', TYPE="Directory", LABEL=os.path.basename(baseDirectoryPath.rstrip('/')))
     structMapDivObjects = createFileSec(os.path.join(baseDirectoryPath, "objects"), structMapDiv)
+    structMapDivObjects.set("ADMID", "amdSec_1")
+    amdSecs.append(create_object_metadata("amdSec_1"))
```
Comment (Hwesta) [inline src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py]: Any particular reason you didn't leave `globalAmdSecCounter` as 0, increment it in `create_object_metadata`, and add the object metadata to the `amdSec` array, similar to how the other amdSecs work?

What if `find_source_metadata` returns two empty arrays? Will an empty amdSec be generated, and is that acceptable?

File: src/MCPClient/lib/clientScripts/archivematicaTransferMetadata.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+sys.path.append("/usr/lib/archivematica/archivematicaCommon")
+import databaseInterface
+
+def fetch_set_uuid(sip_uuid):
+    sql = """SELECT pk FROM TransferMetadataSets
+    INNER JOIN Transfers ON transferMetadataSetRowUUID = pk
+    WHERE transferUUID = '{}'
+    """.format(sip_uuid)
+    print(sql, file=sys.stderr)
```
Comment (Hwesta) [inline src/MCPClient/lib/clientScripts/archivematicaTransferMetadata.py]: Why stderr?  Is the SQL an error?

Comment (mistydemeo) [inline src/MCPClient/lib/clientScripts/archivematicaTransferMetadata.py]: Leftover debug print - shouldn't be here.

File: src/dashboard/src/components/administration/views.py [hunk 9]
```diff
@@ -257,6 +259,46 @@ def api(request):
...
+def term_delete_context(request, term_uuid):
+    term = models.TaxonomyTerm.objects.get(pk=term_uuid)
+    prompt = 'Delete term ' + term.term + '?'
+    cancel_url = reverse("components.administration.views.term_detail", args=[term_uuid])
+    return RequestContext(request, {'action': 'Delete', 'prompt': prompt, 'cancel_url': cancel_url})
+
+@decorators.confirm_required('simple_confirm.html', term_delete_context)
+def term_delete(request, term_uuid):
+    if request.method == 'POST':
+        pass # do delete
```
Comment (Hwesta) [inline src/dashboard/src/components/administration/views.py]: Leftover?

File: src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py (hunk 7 above)
Comment (mistydemeo) [inline src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py]: This hasn't been updated substantially since the last round of createMETS2 changes - I'll update it now to be more consistent.

File: src/dashboard/src/templates/transfer/detail.html [hunk 10]
```diff
@@ -33,11 +33,16 @@
...
+      <h5>Transfer Metadata</h5>
+      <ul>
+        <li><a href="{% url 'components.transfer.views.component' set_uuid %}">Edit</a></li>
+      </ul>
+
       <h5>Metadata</h5>
       <ul>
         <li><a href="{% url 'components.transfer.views.transfer_metadata_list' uuid %}">List</a></li>
         <li><a href="{% url 'components.transfer.views.transfer_metadata_edit' uuid %}">Add</a></li>
-      </ul>
```
Comment (Hwesta) [inline src/dashboard/src/templates/transfer/detail.html:40]: Did you mean to remove this?

File: src/dashboard/src/components/transfer/views.py [hunk 11]
```diff
@@ -66,6 +67,77 @@ def grid(request):
...
+            field.allownewvalue = True
+            # support allownewvalue
+            # by loading taxonomy and checked if it's open
+        try:
+            field_value = models.TransferMetadataFieldValue.objects.get(
+                field=field,
+                set=metadata_set
+            )
+            values[(field.fieldname)] = field_value.fieldvalue
+        except:
```
Comment (Hwesta) [inline src/dashboard/src/components/transfer/views.py]: What exception is being caught?

File: src/dashboard/src/components/transfer/views.py [hunk 12]
```diff
@@ -66,6 +67,77 @@ def grid(request):
...
+            values[(field.fieldname)] = field_value.fieldvalue # override initially loaded value, if any
+
+    if fields_saved:
+        messages.append('Metadata saved.')
+
+    return render(request, 'transfer/component.html', locals())
+
+def browser(request):
+    originals_directory = '/var/archivematica/sharedDirectory/transferBackups/originals'
+    arrange_directory = '/var/archivematica/sharedDirectory/transferBackups/arrange'
```
Comment (Hwesta) [inline src/dashboard/src/components/transfer/views.py]: sharedDirectory should be read from the config file, not hardcoded.  `components.helpers.get_server_config_value('sharedDirectory')` is useful here.

File: src/dashboard/src/components/transfer/views.py [hunk 13]
```diff
@@ -187,3 +260,67 @@ def transfer_metadata_edit(request, uuid, id=None):
...
+        mimetype='application/json'
+    )
+
+def rename_metadata_set(request, set_uuid, placeholder_id):
+    response = {}
+
+    try:
+        path = request.POST.get('path')
+        if not path:
+            raise KeyError
```
Comment (Hwesta) [inline src/dashboard/src/components/transfer/views.py]: Won't `request.POST['path']` raise a KeyError if path isn't found?

File: src/dashboard/src/templates/transfer/detail.html (hunk 10 above)
Comment (mistydemeo) [inline src/dashboard/src/templates/transfer/detail.html:40]: Nope - looks like a rebase victim.

File: src/dashboard/src/components/transfer/views.py (hunk 12 above)
Comment (mistydemeo) [inline src/dashboard/src/components/transfer/views.py]: This actually doesn't look related. The original commit history isn't present in this branch, but is this a commit from an early version of SIP arrange that ended up in the metadata branch maybe?

File: src/dashboard/src/media/js/transfer/component_form.js [hunk 14]
```diff
@@ -40,22 +76,44 @@ var TransferComponentFormView = Backbone.View.extend({
...
+      key = component_keys[i];
+      component = $.extend(true, {}, components[key]);
+      component.path = Base64.encode(component.path);
+      paths.push(component);
+    }
     return paths;
   },
 
   startTransfer: function(transfer) {
+    console.log(transfer.sourcePaths);
```
Comment (Hwesta) [inline src/dashboard/src/media/js/transfer/component_form.js]: Do we want to leave logging in JS?

File: src/dashboard/src/components/transfer/views.py (hunk 12 above)
Comment (mistydemeo) [inline src/dashboard/src/components/transfer/views.py]: Looks like this is an unrelated early SIP arrange function that was previously in the primary branches but never used, and deleted by you in 5a04c9ba0a125db0c82b044e1243477159de56eb. Must have been accidentally restored in the rebase. Will delete.

File: src/dashboard/src/components/administration/forms.py [hunk 15]
```diff
@@ -98,3 +98,11 @@ class ArchivistsToolkitConfigForm(ModelForm):
...
     class Meta:
         model = ArchivistsToolkitConfig
+
+class TaxonomyTermForm(ModelForm):
+    taxonomy = forms.ModelChoiceField(models.Taxonomy.objects.all(), required=True)
+    term = forms.CharField(required=True, widget=TextInput(attrs=settings.INPUT_ATTRS))
+
+    class Meta:
+        model = models.TaxonomyTerm
+        exclude = ('identifiertype')
```
Comment (Hwesta) [inline src/dashboard/src/components/administration/forms.py]: `identifiertype` isn't a field in `models.TaxonomyTerm`

File: src/dashboard/src/components/administration/forms.py [hunk 16]
```diff
@@ -98,3 +98,11 @@ class ArchivistsToolkitConfigForm(ModelForm):
  
     class Meta:
         model = ArchivistsToolkitConfig
+
+class TaxonomyTermForm(ModelForm):
+    taxonomy = forms.ModelChoiceField(models.Taxonomy.objects.all(), required=True)
+    term = forms.CharField(required=True, widget=TextInput(attrs=settings.INPUT_ATTRS))
```
Comment (Hwesta) [inline src/dashboard/src/components/administration/forms.py]: Do these fields need to be defined in the form?  They're both in the model, and should be created automatically.  I think if a ForeignKey is not NULLable, then it is required in the form.  To set widget specifics for a field, use https://docs.djangoproject.com/en/dev/topics/forms/modelforms/#overriding-the-default-fields

File: src/dashboard/src/components/transfer/views.py (hunk 11 above)
Comment (mistydemeo) [inline src/dashboard/src/components/transfer/views.py]: I'm guessing it's `DoesNotExist`, so updated to specify.

File: src/dashboard/src/components/transfer/views.py (hunk 12 above)
Comment (Hwesta) [inline src/dashboard/src/components/transfer/views.py]: Ah, okay. I'm still wrapping my head around the feature, so I can't tell what's actually used and what isn't. :)

File: src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py (hunk 7 above)
Comment (Hwesta) [inline src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py]: Nice fix.

File: src/dashboard/src/components/administration/forms.py (hunk 15 above)
Comment (mistydemeo) [inline src/dashboard/src/components/administration/forms.py]: Nor was it in the original SQL table, so not sure why that was specified. I'll remove.

File: src/dashboard/src/components/administration/forms.py (hunk 16 above)
Comment (mistydemeo) [inline src/dashboard/src/components/administration/forms.py]: Thanks! Updated.

File: src/dashboard/src/main/models.py [hunk 17]
```diff
@@ -626,3 +627,60 @@ class AtkDIPObjectResourcePairing(models.Model):
...
+    id = UUIDPkField()
+    createdtime = models.DateTimeField(db_column='createdTime', auto_now_add=True)
+    set = models.ForeignKey('TransferMetadataSet', db_column='setUUID', to_field='id')
+    field = models.ForeignKey('TransferMetadataField', db_column='fieldUUID', to_field='id')
+    fieldvalue = models.TextField(blank=True, db_column='fieldValue')
+
+    class Meta:
+        db_table = u'TransferMetadataFieldValues'
+
+class Taxonomy(models.Model):
```
Comment (Hwesta) [inline src/dashboard/src/main/models.py:668]: Can you leave a comment in this area explaining why the field definitions are done in as separate tables, instead of fields of `Taxonomy`?  Flexibility, want to edit later, etc.

Comment (mistydemeo) [inline src/dashboard/src/main/models.py:668]: Added comment above.

File: src/dashboard/src/components/administration/urls.py [hunk 18]
```diff
@@ -32,4 +33,9 @@
     (r'api/$', 'api'),
     (r'general/$', 'general'),
     (r'version/$', 'version'),
+    (r'taxonomy/term/(?P<term_uuid>' + settings.UUID_REGEX + ')/$', 'term_detail'),
+    (r'taxonomy/term/(?P<term_uuid>' + settings.UUID_REGEX + ')/delete/$', 'term_delete'),
+    (r'taxonomy/(?P<taxonomy_uuid>' + settings.UUID_REGEX + ')/$', 'terms'),
+    (r'taxonomy/$', 'taxonomy'),
+    (r'general/$', 'general'),
```
Comment (Hwesta) [inline src/dashboard/src/components/administration/urls.py]: Looks like a rebase dupe.

File: src/dashboard/src/templates/administration/taxonomy.html [hunk 19]
```diff
@@ -0,0 +1,47 @@
+{% extends "layout_fluid.html" %}
+{% load breadcumbs %}
```
Comment (Hwesta) [inline src/dashboard/src/templates/administration/taxonomy.html]: This should be `{% load breadcrumb %}`, as well as for term_detail.html & terms.html

**Comment by Hwesta** [comment]
I'm out of things to comment on. :P Looks good to merge.
